tests/unit/locking_test.cc

```diff
@@ -420,3 +422,170 @@ SEASTAR_THREAD_TEST_CASE(test_with_lock_typed_return_throwing_move_func) {
+
+    {
+        const auto ulock = co_await get_unique_lock(sm);
+        BOOST_REQUIRE(!sm.try_lock());
+        BOOST_REQUIRE(!sm.try_lock_shared());
+    }
+
+    BOOST_REQUIRE(sm.try_lock());
+    sm.unlock();
+
+    {
+        const auto slock1 = co_await get_shared_lock(sm);
+        BOOST_REQUIRE(!sm.try_lock());
+        BOOST_REQUIRE(sm.try_lock_shared());
+
+        const auto slock2 = co_await get_shared_lock(sm);
+
+        // This balances out the call to `try_lock_shared()` above.
+        sm.unlock_shared();
+    }
+
+    BOOST_REQUIRE(sm.try_lock());
+    sm.unlock();
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_exclusive_locks) {
+    shared_mutex sm{};
+    unsigned counter = 0;
+
+    co_await coroutine::parallel_for_each(boost::irange(0, 10), coroutine::lambda([&sm, &counter] (auto&&) -> future<> {
+        const auto ulock = co_await get_unique_lock(sm);
+        BOOST_REQUIRE_EQUAL(counter, 0u);
+        ++counter;
+        co_await sleep(1ms);
+        --counter;
+        BOOST_REQUIRE_EQUAL(counter, 0u);
+    }));
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_shared_locks) {
+    shared_mutex sm{};
+    unsigned counter = 0;
+    unsigned max = 0;
+
+    co_await coroutine::parallel_for_each(boost::irange(0, 10),
+            coroutine::lambda([&sm, &counter, &max] (auto&&) -> future<> {
+        const auto slock = co_await get_shared_lock(sm);
+        ++counter;
+        max = std::max(max, counter);
+        co_await sleep(1ms);
+        --counter;
+    }));
+
+    BOOST_REQUIRE_EQUAL(counter, 0u);
+    BOOST_REQUIRE_NE(max, 0u);
+}
+
+
+SEASTAR_TEST_CASE(test_shared_mutex_failed_func_locks) {
+    shared_mutex sm;
+
+    // verify that the shared_mutex is unlocked when func fails
+    try {
+        const auto slock = co_await get_shared_lock(sm);
+        throw std::runtime_error("injected");
+    } catch (const std::runtime_error&) {
+        BOOST_REQUIRE(sm.try_lock());
+        sm.unlock();
+    } catch (...) {
+        BOOST_FAIL(format("Unexpected exception type: {}", std::current_exception()));
+    }
+
+    try {
+        const auto ulock = co_await get_unique_lock(sm);
+        throw std::runtime_error("injected");
+    } catch (const std::runtime_error&) {
+        BOOST_REQUIRE(sm.try_lock());
+        sm.unlock();
+    } catch (...) {
+        BOOST_FAIL(format("Unexpected exception type: {}", std::current_exception()));
+    }
+
+    BOOST_REQUIRE(sm.try_lock());
+    sm.unlock();
+}
+
+SEASTAR_TEST_CASE(test_shared_mutex_failed_lock_locks) {
+#ifdef SEASTAR_ENABLE_ALLOC_FAILURE_INJECTION
+    shared_mutex sm;
+
+    // If l.lock() fails, neither the function, nor l.unlock()
+    // should be called.
+    const auto ulock = co_await get_unique_lock(sm);
+
+    seastar::memory::local_failure_injector().fail_after(0);
+    BOOST_REQUIRE_THROW(co_await std::invoke([&] () -> future<> {
+        const auto slock = co_await get_shared_lock(sm);
+        BOOST_REQUIRE(false);
+    }), std::bad_alloc);
```

Comment:
Can you please help me understand what we're trying to test here? That the coroutine functions need to allocate and if they can't, they fail before even starting? Isn't this obvious for all coroutine functions? Is this really what you hoped to test here?

Reply:
As in https://github.com/scylladb/seastar/pull/2196#discussion_r1606063265, this test is an analogue of the existing one, adjusted to using locks and coroutinised. The way I understand the logic here is we want to check what happens in case of an allocation failure: we enter a coroutine and await a lock. However, the lock will never be returned because the allocation of the coroutine awaiting it will fail, throwing an `std::bad_alloc` (alternatively, the underlying allocation of a promise performed by `seastar::shared_mutex` will fail. Either way, there should be an allocation failure). However, now that I look at it, it seems overly complex and probably unnecessary; chances are I also don't understand it correctly. I can remove the test.